sematic/calculator.py

```diff
@@ -61,6 +60,62 @@ def __init__(
         self.__doc__ = func.__doc__
         self.__module__ = func.__module__
         self.__name__ = func.__name__
+        self._validate()
+
+    def _validate(self):
+        for key, annotation in self._input_types.items():
+            try:
+                validate_type_annotation(annotation)
+            except TypeError as e:
+                raise TypeError(
+                    f"Invalid type annotation for argument '{key}' of {repr(self)}: {e}"
+                )
+        try:
+            validate_type_annotation(self._output_type)
+        except TypeError as e:
+            raise TypeError(f"Invalid type annotation for output of {repr(self)}: {e}")
+
+    @classmethod
+    def _validate_func(cls, func):
+        is_method_like = (
+            inspect.ismethod(func)
+            or
+            # Why is this second check necessary? Because if somebody uses @func
+            # on a method inside the class itself, the method is not yet bound
+            # to the class, and inspect's ismethod won't pick it up.
+            (
+                hasattr(func, "__call__")
+                and len(inspect.signature(func).parameters.keys()) >= 1
+                and list(inspect.signature(func).parameters.keys())[0]
+                in {"self", "cls"}
+            )
+        )
+        if is_method_like:
+            raise TypeError(
+                "@sematic.func can only be used with functions, not methods. "
+                f"But {repr(func)} is a method. Instead, consider defining a "
+                "function where the first argument is the object currently being "
+                "passed as 'self' or 'cls'."
```

Comment:
My understanding is that even a function with the first parameter named `self` or `cls` will be considered a method according to the testing code.

I'm concerned that a user initially writing a method will see the error message, hastily extract the method out to a function and try it before renaming the first parameter, and obtain the same error message, which would be quite frustrating. We should increase transparency into how this check is done, or how to do the extraction.

```suggestion
                "passed as 'self' or 'cls', renamed to something more descriptive."
```

Reply:
changed the wording, as suggested